### PR TITLE
JAVAFICATION: Port memory queue read client to Java. …

### PR DESCRIPTION
**Pull request by danhermann**
Also uses Java counters more efficiently per Armin's work in #8985. Rough throughput testing with generator input shows 15-20% increase due mainly to the metrics improvements.

### CODE REVIEW

File: logstash-core/src/main/java/org/logstash/ext/JrubyMemoryReadBatchExt.java [hunk 1]
```diff
@@ -11,20 +11,32 @@
...
 import org.jruby.runtime.ThreadContext;
 import org.jruby.runtime.builtin.IRubyObject;
+import org.logstash.RubyUtil;
 
 @JRubyClass(name = "MemoryReadBatch")
 public class JrubyMemoryReadBatchExt extends RubyObject {
 
     private LinkedHashSet<IRubyObject> events;
 
     public JrubyMemoryReadBatchExt(final Ruby runtime, final RubyClass metaClass) {
```
Comment (danhermann) [inline logstash-core/src/main/java/org/logstash/ext/JrubyMemoryReadBatchExt.java]: Had to keep this constructor due to the static setup performed in the `RubyUtil` class.

**Comment by danhermann** [comment]
@original-brownbear, can you look this one over when you have a chance?

**Comment by original-brownbear** [comment]
@danhermann sure thing will take a look shortly :)

**Comment by danhermann** [comment]
Jenkins test this please

File: logstash-core/src/main/java/org/logstash/ext/JrubyMemoryReadBatchExt.java (hunk 1 above)
Comment (original-brownbear) [reply logstash-core/src/main/java/org/logstash/ext/JrubyMemoryReadBatchExt.java]: @danhermann Right ... my bad :)

**Comment by original-brownbear** [comment]
@danhermann LGTM, feel free to squash + merge :)

**Comment by original-brownbear** [comment]
@danhermann nice one, I benchmarked this as well now and get an improvement from `380k/s` (master) to `420k/s` with this branch for:

```sh
 bin/logstash -b 128 -w 1 -e 'input { generator {threads => 1}} output {stdout {codec => "dots"}}'  | pv | wc -c
```

File: logstash-core/src/main/java/org/logstash/ext/JrubyMemoryReadClientExt.java [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+            long elapsedTimeMillis = (System.nanoTime() - startTime) / 1_000_000;
+            eventMetricTime.increment(elapsedTimeMillis);
+            pipelineMetricTime.increment(elapsedTimeMillis);
+        }
+        return this;
+    }
+
+    @JRubyMethod(name = "add_filtered_metrics", required = 1)
+    public IRubyObject addFilteredMetrics(final ThreadContext context, IRubyObject filteredSize) {
+        eventMetricFiltered.increment(((RubyNumeric) filteredSize).getIntValue());
```
Comment (original-brownbear) [inline logstash-core/src/main/java/org/logstash/ext/JrubyMemoryReadClientExt.java]: @danhermann 

small improvement if you have a sec :) Just cache:

```java
(RubyNumeric) filteredSize).getIntValue()
```

to a variable :)

File: logstash-core/src/main/java/org/logstash/ext/JrubyMemoryReadClientExt.java [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+        pipelineMetricFiltered.increment(((RubyNumeric) filteredSize).getIntValue());
+        return this;
+    }
+
+    @JRubyMethod(name = "add_output_metrics", required = 1)
+    public IRubyObject addOutputMetrics(final ThreadContext context, IRubyObject filteredSize) {
+        eventMetricOut.increment(((RubyNumeric) filteredSize).getIntValue());
```
Comment (original-brownbear) [inline logstash-core/src/main/java/org/logstash/ext/JrubyMemoryReadClientExt.java]: @danhermann 

small improvement if you have a sec :) Just cache:

```java
(RubyNumeric) filteredSize).getIntValue()
```

to a variable :)

File: logstash-core/src/main/java/org/logstash/ext/JrubyMemoryReadClientExt.java [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+    }
+
+    @JRubyMethod(name = "start_metrics")
+    public IRubyObject rubyStartMetrics(final ThreadContext context, IRubyObject batch) {
+        startMetrics((JrubyMemoryReadBatchExt) batch);
+        return this;
+    }
+
+    private void startMetrics(JrubyMemoryReadBatchExt batch) {
+        inflightBatches.put(Thread.currentThread().getId(), batch);
```
Comment (original-brownbear) [inline logstash-core/src/main/java/org/logstash/ext/JrubyMemoryReadClientExt.java]: @danhermann 

small improvement if you have a sec :) Just cache:

```java
Thread.currentThread().getId()
```

to a variable :)

**Comment by original-brownbear** [comment]
@danhermann looks just fine, failure is known #8142 :)

**Comment by danhermann** [comment]
jenkins test this please

**Pull request closed without merging**